pkg/kudoctl/cmd/diagnostics/diagnostics.go

```diff
@@ -0,0 +1,105 @@
+package diagnostics
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/kudobuilder/kudo/pkg/kudoctl/env"
+	"github.com/kudobuilder/kudo/pkg/kudoctl/kube"
+	"github.com/kudobuilder/kudo/pkg/version"
+	"github.com/spf13/cobra"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const (
+	nsKudoSystem      = "kudo-system"
+	labelKudoOperator = "kudo.dev/operator"
+	appKudoManager    = "kudo-manager"
+)
+
+var Options = struct {
+	Instance string
+}{}
+
+type Collector interface {
+	Collect(f writerFactory) error
+}
+
+func Collect(cmd *cobra.Command, settings *env.Settings) error {
+	fmt.Println("Collecting diagnostics")
+
+	config, _ := clientcmd.BuildConfigFromFlags("", settings.KubeConfig)
+	ns := settings.Namespace
+	kc, _ := env.GetClient(settings)
+	instance, _ := kc.GetInstance(Options.Instance, ns)
+	c, _ := kube.GetKubeClient(settings.KubeConfig)
+
+	byOperator := metav1.ListOptions{
+		LabelSelector: labelKudoOperator + "=" + instance.Labels[labelKudoOperator],
+	}
+	byKudoManager := metav1.ListOptions{
+		LabelSelector: "app=" + appKudoManager,
+	}
+
+	iResources := &resourceFuncs{c, kc, ns, byOperator, instance}
+	cResources := &resourceFuncs{c, nil, nsKudoSystem, byKudoManager, nil} // no need for kudo and instance
+
+	// TODO: use more meaningful variable names
+	dc := resourceListCollector{getResources: iResources.deployments()}
+	pc := resourceListCollector{getResources: iResources.pods()}
+	ec := resourceListCollector{getResources: iResources.events()}
+	sc := resourceListCollector{getResources: iResources.services()}
+	rsc := resourceListCollector{getResources: iResources.replicaSets()}
+	ovc := resourceCollector{getResource: iResources.operatorVersions()}
+	oc := resourceCollector{getResource: iResources.operators(&ovc)}
+	lc := logCollector{
+		Client: c,
+		ns:     ns,
+		opts:   corev1.PodLogOptions{}, // TODO: add time range
+		logs:   make(map[string]io.ReadCloser),
+		pods:   &pc,
+	}
+	cpc := resourceListCollector{getResources: cResources.pods()}
+	ssc := resourceListCollector{getResources: cResources.statefulSets()}
+	rbc := resourceListCollector{getResources: cResources.roleBindings()}
+	crbc := resourceListCollector{getResources: cResources.clusterRoleBindings()}
+	rc := resourceListCollector{getResources: cResources.roles(&rbc)}
```

Comment:
I looked at a couple of these functions and it does not look like they use the arguments.
If they are independent then you can do without intermediate variables and just replace this with a series of:
```
collectors = append(collectors, resourceListCollector{...})
```
or such.